cdap-watchdog/src/main/java/co/cask/cdap/logging/appender/LogPublisher.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * Copyright © 2018 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.appender;
+
+import ch.qos.logback.core.spi.ContextAware;
+import co.cask.cdap.common.service.AbstractRetryableScheduledService;
+import co.cask.cdap.common.service.RetryStrategy;
+import kafka.producer.KeyedMessage;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.concurrent.BlockingQueue;
+
+/**
+ * A class that continuously serializes logs from a queue and publishes them.
+ *
+ * @param <BUFFER_MESSAGE> the type of message used in the in-memory buffer, before publishing
+ */
+public abstract class LogPublisher<BUFFER_MESSAGE> extends AbstractRetryableScheduledService {
+
+  private final int queueSize;
+  private final BlockingQueue<LogMessage> messageQueue;
+  private final ContextAware contextAware;
+  private final List<BUFFER_MESSAGE> buffer;
+  private volatile boolean hadFailure;
+  private volatile Thread blockingThread;
+
+  public LogPublisher(int queueSize, BlockingQueue<LogMessage> messageQueue, RetryStrategy retryStrategy,
+                      ContextAware contextAware) {
+    super(retryStrategy);
+    this.queueSize = queueSize;
+    this.messageQueue = messageQueue;
+    this.contextAware = contextAware;
+    this.buffer = new ArrayList<>(queueSize);
+  }
+
+  @Override
+  protected long runTask() throws Exception {
+    // Only block for messages if it is not a failure retry
+    publishMessages(buffer, hadFailure);
+    // We only clear the buffer once the messages are successfully published
+    buffer.clear();
+    hadFailure = false;
+    return 0;
+  }
+
+  @Override
+  protected long handleRetriesExhausted(Exception e) {
+    // Log using the status manager
+    contextAware.addError("Failed to publish log message by " + getServiceName(), e);
+    return 0;
+  }
+
+    @Override
+  protected boolean shouldRetry(Exception ex) {
+    hadFailure = true;
+    return super.shouldRetry(ex);
+  }
+
+    @Override
+  protected void doShutdown() throws Exception {
+    // Interrupt the run thread first
+    // If the run loop is sleeping / blocking, it will wake and break the loop
+    Thread runThread = this.blockingThread;
+    if (runThread != null) {
+      runThread.interrupt();
+    }
+
+    // Publish all remaining messages.
```

Comment:
You should clear the interrupt of the current thread by calling `Thread.interrupted()`